clang/lib/Sema/SemaExpr.cpp

```diff
@@ -17600,6 +17616,184 @@ ExprResult Sema::BuildSourceLocExpr(SourceLocIdentKind Kind, QualType ResultTy,
       SourceLocExpr(Context, Kind, ResultTy, BuiltinLoc, RPLoc, ParentContext);
 }
 
+ExprResult Sema::ActOnPPEmbedExpr(SourceLocation BuiltinLoc,
+                                  SourceLocation BinaryDataLoc,
+                                  SourceLocation RPLoc, StringLiteral *Filename,
+                                  StringLiteral *BinaryData) {
+  return new (Context)
+      PPEmbedExpr(Context, Filename, BinaryData, BuiltinLoc, RPLoc, CurContext);
+}
+
+IntegerLiteral *Sema::ExpandSinglePPEmbedExpr(PPEmbedExpr *PPEmbed,
+                                              bool FirstElement) {
+  assert((PPEmbed->getDataElementCount(Context) == 1 || !FirstElement) &&
+         "Data should only contain a single element");
+  StringLiteral *DataLiteral = PPEmbed->getDataStringLiteral();
+  QualType ElementTy = PPEmbed->getType();
+  const size_t TargetWidth = Context.getTypeSize(ElementTy);
+  const size_t BytesPerElement = CHAR_BIT / TargetWidth;
+  StringRef Data = DataLiteral->getBytes();
+  Data = Data.substr(FirstElement ? 0 : Data.size() - 1, 1);
+  SmallVector<uint64_t, 4> ByteVals{};
+  for (size_t ValIndex = 0; ValIndex < BytesPerElement; ++ValIndex) {
+    if ((ValIndex % sizeof(uint64_t)) == 0) {
+      ByteVals.push_back(0);
+    }
+    const unsigned char DataByte = Data[ValIndex];
```

Comment:
We look at the type of elements but them assume they are unsigned char?

Reply:
Hmmm actually, I don't think we need this complication at all, do we? We don't support any non-8-bit char architectures currently, so we should be able to assume that the byte we read in is the byte we emit, yes?